### PR TITLE
[SPARK-42780][BUILD] Upgrade `Tink` to 1.9.0

### PR DESCRIPTION
**Pull request by bjornjorgensen**
### What changes were proposed in this pull request?
Upgrade google Tink from 1.7.0 to 1.9.0

[Release note for 1.8.0](https://github.com/tink-crypto/tink-java/releases/tag/v1.8.0)

[Release note for 1.9.0](https://github.com/tink-crypto/tink-java/releases/tag/v1.9.0)

### Why are the changes needed?
[SNYK-JAVA-COMGOOGLEPROTOBUF-3040284](https://security.snyk.io/vuln/SNYK-JAVA-COMGOOGLEPROTOBUF-3040284)

[SNYK-JAVA-COMGOOGLEPROTOBUF-3167772](https://security.snyk.io/vuln/SNYK-JAVA-COMGOOGLEPROTOBUF-3167772)

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
This have be benchmarks tested 
With
"com.google.crypto.tink" % "tink" % "1.6.1"
(min, avg, max) = (75024163.500, 76331532.832, 77324718.069), stdev = 652319.870

With
"com.google.crypto.tink" % "tink" % "1.9.0"
(min, avg, max) = (76279051.841, 77512667.749, 78590966.453), stdev = 632832.384

Almost the same.. Think 1.9.0 is perhaps a bit slower.

Pass GA

### CODE REVIEW

**Comment by bjornjorgensen** [comment]
https://github.com/apache/spark/pull/40408 is related to this. 
CC @LuciferYang

File: dev/deps/spark-deps-hadoop-3-hive-2.3 [hunk 1]
```diff
@@ -59,9 +59,12 @@ datanucleus-core/4.1.17//datanucleus-core-4.1.17.jar
 datanucleus-rdbms/4.1.19//datanucleus-rdbms-4.1.19.jar
 derby/10.14.2.0//derby-10.14.2.0.jar
 dropwizard-metrics-hadoop-metrics2-reporter/0.1.2//dropwizard-metrics-hadoop-metrics2-reporter-0.1.2.jar
+error_prone_annotations/2.18.0//error_prone_annotations-2.18.0.jar
 flatbuffers-java/1.12.0//flatbuffers-java-1.12.0.jar
 gcs-connector/hadoop3-2.2.11/shaded/gcs-connector-hadoop3-2.2.11-shaded.jar
 gmetric4j/1.0.10//gmetric4j-1.0.10.jar
+google-http-client/1.43.1//google-http-client-1.43.1.jar
+grpc-context/1.27.2//grpc-context-1.27.2.jar
```
Comment (LuciferYang) [inline dev/deps/spark-deps-hadoop-3-hive-2.3:67]: Why does this version have GRPC related dependencies? Can we exclude them.

CC @HyukjinKwon FYI

Comment (LuciferYang) [reply dev/deps/spark-deps-hadoop-3-hive-2.3:67]: We just used 

```
import com.google.crypto.tink.subtle.AesGcmJce;
import com.google.crypto.tink.subtle.Hkdf;
import com.google.crypto.tink.subtle.Random;
import com.google.crypto.tink.subtle.X25519;
```
in `AuthEngine`, so let's exclude 

```
<exclusions>
        <exclusion>
          <groupId>com.google.errorprone</groupId>
          <artifactId>error_prone_annotations</artifactId>
        </exclusion>
        <exclusion>
          <groupId>com.google.http-client</groupId>
          <artifactId>google-http-client</artifactId>
        </exclusion>
      </exclusions>
```
from `tink` as default

File: pom.xml [hunk 2]
```diff
@@ -399,6 +399,21 @@
       <artifactId>unused</artifactId>
       <version>1.0.0</version>
     </dependency>
+    <dependency>
```
Comment (bjornjorgensen) [inline pom.xml:402]: Is this the right place?

Comment (LuciferYang) [reply pom.xml:402]: https://github.com/apache/spark/blob/a28d2d7e900e8e2c65953fddbf32708802c4a700/pom.xml#L2691-L2695

I think we should add `exclusions` here

**Comment by bjornjorgensen** [comment]
@LuciferYang Thank you 👍

**Comment by bjornjorgensen** [comment]
@HyukjinKwon will you have a look at this?

**Review by dongjoon-hyun — COMMENTED** [review]
If you don't mind, could you update the PR description with a micro-benchmark like #35679 ?

**Comment by bjornjorgensen** [comment]
> If you don't mind, could you update the PR description with a micro-benchmark like #35679 ?

#35679 is a PR for "There is a performance optimization ..."
This PR is only to fix dependents that did have CVE issues. 

I dont think that benchmarks for this PR is not relevant.

**Comment by dongjoon-hyun** [comment]
I believe we need to see that this PR doesn't cause any perf regression, @bjornjorgensen .

**Comment by bjornjorgensen** [comment]
Oh.. I think you have found this https://github.com/tink-crypto/tink-java/issues/6 
He is using 1.7.0 
https://github.com/ks-yim/tink-bm/blob/960a88d69934b40dcfdd9b2aeb94028c64393a84/build.gradle.kts#LL27C11-L27C11
Witch we are also using https://github.com/apache/spark/pull/37473

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
@bjornjorgensen . It's a simple validation, isn't it?

Without guaranteeing that there is no side-effect in perf, I'm negative with this upgrade,

**Comment by bjornjorgensen** [comment]
Here https://github.com/bjornjorgensen/GoogleThinkBenchmark is the code that I have used to test it.

"com.google.crypto.tink" % "tink" % "1.6.1"
```
[info] compiling 5 Java sources to /home/bjorn/Documents/github/GoogleThinkBenchmark/target/scala-2.12/classes ...
[info] running (fork) org.openjdk.jmh.Main 
[info] # JMH version: 1.31
[info] # VM version: JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10-Ubuntu-1ubuntu2
[info] # VM invoker: /usr/lib/jvm/java-17-openjdk-amd64/bin/java
[info] # VM options: <none>
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 5 iterations, 10 s each
[info] # Measurement: 5 iterations, 10 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt
[info] # Run progress: 0.00% complete, ETA 00:08:20
[info] # Fork: 1 of 5
[info] # Warmup Iteration   1: 79207785.709 ns/op
[info] # Warmup Iteration   2: 75605599.346 ns/op
[info] # Warmup Iteration   3: 76214904.538 ns/op
[info] # Warmup Iteration   4: 76989327.792 ns/op
[info] # Warmup Iteration   5: 77099889.177 ns/op
[info] Iteration   1: 75762018.553 ns/op
[info] Iteration   2: 75379888.865 ns/op
[info] Iteration   3: 75498314.789 ns/op
[info] Iteration   4: 75024163.500 ns/op
[info] Iteration   5: 75680816.338 ns/op
[info] # Run progress: 20.00% complete, ETA 00:06:43
[info] # Fork: 2 of 5
[info] # Warmup Iteration   1: 80245617.368 ns/op
[info] # Warmup Iteration   2: 75652339.774 ns/op
[info] # Warmup Iteration   3: 77061164.331 ns/op
[info] # Warmup Iteration   4: 76287806.765 ns/op
[info] # Warmup Iteration   5: 75317346.985 ns/op
[info] Iteration   1: 75686301.880 ns/op
[info] Iteration   2: 75330525.474 ns/op
[info] Iteration   3: 75734112.594 ns/op
[info] Iteration   4: 76288954.348 ns/op
[info] Iteration   5: 76606136.962 ns/op
[info] # Run progress: 40.00% complete, ETA 00:05:02
[info] # Fork: 3 of 5
[info] # Warmup Iteration   1: 80973643.427 ns/op
[info] # Warmup Iteration   2: 77317109.354 ns/op
[info] # Warmup Iteration   3: 76322130.439 ns/op
[info] # Warmup Iteration   4: 76301836.621 ns/op
[info] # Warmup Iteration   5: 77046937.308 ns/op
[info] Iteration   1: 76474108.687 ns/op
[info] Iteration   2: 77131903.554 ns/op
[info] Iteration   3: 76728218.573 ns/op
[info] Iteration   4: 77324718.069 ns/op
[info] Iteration   5: 76930795.244 ns/op
[info] # Run progress: 60.00% complete, ETA 00:03:21
[info] # Fork: 4 of 5
[info] # Warmup Iteration   1: 80497824.184 ns/op
[info] # Warmup Iteration   2: 76670359.000 ns/op
[info] # Warmup Iteration   3: 77410211.954 ns/op
[info] # Warmup Iteration   4: 76880883.221 ns/op
[info] # Warmup Iteration   5: 76903019.008 ns/op
[info] Iteration   1: 76878110.405 ns/op
[info] Iteration   2: 77073021.462 ns/op
[info] Iteration   3: 76818391.687 ns/op
[info] Iteration   4: 76985174.577 ns/op
[info] Iteration   5: 76645837.389 ns/op
[info] # Run progress: 80.00% complete, ETA 00:01:40
[info] # Fork: 5 of 5
[info] # Warmup Iteration   1: 80435216.312 ns/op
[info] # Warmup Iteration   2: 76728526.878 ns/op
[info] # Warmup Iteration   3: 76168514.106 ns/op
[info] # Warmup Iteration   4: 75383839.180 ns/op
[info] # Warmup Iteration   5: 75929712.985 ns/op
[info] Iteration   1: 76427345.038 ns/op
[info] Iteration   2: 76718093.679 ns/op
[info] Iteration   3: 76872120.244 ns/op
[info] Iteration   4: 76085370.545 ns/op
[info] Iteration   5: 76203878.348 ns/op
[info] Result "com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt":
[info]   76331532.832 ±(99.9%) 488639.588 ns/op [Average]
[info]   (min, avg, max) = (75024163.500, 76331532.832, 77324718.069), stdev = 652319.870
[info]   CI (99.9%): [75842893.244, 76820172.420] (assumes normal distribution)
[info] # Run complete. Total time: 00:08:24
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                  Mode  Cnt         Score        Error  Units
[info] GoogleThinkBenchmark.testAesGcmJceEncrypt  avgt   25  76331532.832 ± 488639.588  ns/op
[success] Total time: 509 s (08:29), completed May 7, 2023, 7:19:58 PM
```

"com.google.crypto.tink" % "tink" % "1.9.0"
```
[info] compiling 5 Java sources to /home/bjorn/Documents/github/GoogleThinkBenchmark/target/scala-2.12/classes ...
[info] running (fork) org.openjdk.jmh.Main 
[info] # JMH version: 1.31
[info] # VM version: JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10-Ubuntu-1ubuntu2
[info] # VM invoker: /usr/lib/jvm/java-17-openjdk-amd64/bin/java
[info] # VM options: <none>
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 5 iterations, 10 s each
[info] # Measurement: 5 iterations, 10 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt
[info] # Run progress: 0.00% complete, ETA 00:08:20
[info] # Fork: 1 of 5
[info] # Warmup Iteration   1: 81974464.393 ns/op
[info] # Warmup Iteration   2: 78344967.031 ns/op
[info] # Warmup Iteration   3: 77870422.233 ns/op
[info] # Warmup Iteration   4: 76366142.015 ns/op
[info] # Warmup Iteration   5: 76969131.346 ns/op
[info] Iteration   1: 77587424.279 ns/op
[info] Iteration   2: 77137913.315 ns/op
[info] Iteration   3: 76307498.705 ns/op
[info] Iteration   4: 76561784.282 ns/op
[info] Iteration   5: 77806261.512 ns/op
[info] # Run progress: 20.00% complete, ETA 00:06:42
[info] # Fork: 2 of 5
[info] # Warmup Iteration   1: 80478869.944 ns/op
[info] # Warmup Iteration   2: 77009168.477 ns/op
[info] # Warmup Iteration   3: 77253208.562 ns/op
[info] # Warmup Iteration   4: 77522236.047 ns/op
[info] # Warmup Iteration   5: 77473338.262 ns/op
[info] Iteration   1: 77294234.708 ns/op
[info] Iteration   2: 77547845.969 ns/op
[info] Iteration   3: 77319105.646 ns/op
[info] Iteration   4: 77437665.792 ns/op
[info] Iteration   5: 76279051.841 ns/op
[info] # Run progress: 40.00% complete, ETA 00:05:02
[info] # Fork: 3 of 5
[info] # Warmup Iteration   1: 81151850.952 ns/op
[info] # Warmup Iteration   2: 77319736.377 ns/op
[info] # Warmup Iteration   3: 77127278.746 ns/op
[info] # Warmup Iteration   4: 77566176.907 ns/op
[info] # Warmup Iteration   5: 77195342.962 ns/op
[info] Iteration   1: 77081135.108 ns/op
[info] Iteration   2: 77275550.108 ns/op
[info] Iteration   3: 76869123.466 ns/op
[info] Iteration   4: 77654387.488 ns/op
[info] Iteration   5: 77896922.171 ns/op
[info] # Run progress: 60.00% complete, ETA 00:03:21
[info] # Fork: 4 of 5
[info] # Warmup Iteration   1: 82232457.566 ns/op
[info] # Warmup Iteration   2: 78139230.227 ns/op
[info] # Warmup Iteration   3: 78621229.438 ns/op
[info] # Warmup Iteration   4: 78189807.516 ns/op
[info] # Warmup Iteration   5: 78500250.758 ns/op
[info] Iteration   1: 78417104.656 ns/op
[info] Iteration   2: 77769043.891 ns/op
[info] Iteration   3: 78550508.242 ns/op
[info] Iteration   4: 78590966.453 ns/op
[info] Iteration   5: 78112666.589 ns/op
[info] # Run progress: 80.00% complete, ETA 00:01:40
[info] # Fork: 5 of 5
[info] # Warmup Iteration   1: 80581520.376 ns/op
[info] # Warmup Iteration   2: 77263616.777 ns/op
[info] # Warmup Iteration   3: 76836909.809 ns/op
[info] # Warmup Iteration   4: 77991064.147 ns/op
[info] # Warmup Iteration   5: 76437476.748 ns/op
[info] Iteration   1: 77256809.777 ns/op
[info] Iteration   2: 78291977.688 ns/op
[info] Iteration   3: 77834381.442 ns/op
[info] Iteration   4: 77869096.798 ns/op
[info] Iteration   5: 77068233.792 ns/op
[info] Result "com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt":
[info]   77512667.749 ±(99.9%) 474041.907 ns/op [Average]
[info]   (min, avg, max) = (76279051.841, 77512667.749, 78590966.453), stdev = 632832.384
[info]   CI (99.9%): [77038625.841, 77986709.656] (assumes normal distribution)
[info] # Run complete. Total time: 00:08:23
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                  Mode  Cnt         Score        Error  Units
[info] GoogleThinkBenchmark.testAesGcmJceEncrypt  avgt   25  77512667.749 ± 474041.907  ns/op
[success] Total time: 510 s (08:30), completed May 7, 2023, 7:34:13 PM
```

**Comment by bjornjorgensen** [comment]
Reboot and run with `sbt 'jmh:run -i 10 -wi 10 -f 1 -t 1'`
with
"com.google.crypto.tink" % "tink" % "1.9.0"
```
[info] compiling 5 Java sources to /home/bjorn/Documents/github/GoogleThinkBenchmark/target/scala-2.12/classes ...
[info] running (fork) org.openjdk.jmh.Main -i 10 -wi 10 -f 1 -t 1
[info] # JMH version: 1.31
[info] # VM version: JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10-Ubuntu-1ubuntu2
[info] # VM invoker: /usr/lib/jvm/java-17-openjdk-amd64/bin/java
[info] # VM options: <none>
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 10 iterations, 10 s each
[info] # Measurement: 10 iterations, 10 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt
[info] # Run progress: 0.00% complete, ETA 00:03:20
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 83028380.132 ns/op
[info] # Warmup Iteration   2: 78802359.000 ns/op
[info] # Warmup Iteration   3: 79210527.969 ns/op
[info] # Warmup Iteration   4: 78695687.141 ns/op
[info] # Warmup Iteration   5: 78938831.701 ns/op
[info] # Warmup Iteration   6: 78575255.695 ns/op
[info] # Warmup Iteration   7: 79112203.724 ns/op
[info] # Warmup Iteration   8: 79323400.079 ns/op
[info] # Warmup Iteration   9: 78537590.867 ns/op
[info] # Warmup Iteration  10: 78413295.234 ns/op
[info] Iteration   1: 77933768.721 ns/op
[info] Iteration   2: 79192640.055 ns/op
[info] Iteration   3: 79182903.512 ns/op
[info] Iteration   4: 79111605.283 ns/op
[info] Iteration   5: 78859743.535 ns/op
[info] Iteration   6: 78318937.016 ns/op
[info] Iteration   7: 77986910.411 ns/op
[info] Iteration   8: 78648235.734 ns/op
[info] Iteration   9: 78609703.445 ns/op
[info] Iteration  10: 79423429.968 ns/op
[info] Result "com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt":
[info]   78726787.768 ±(99.9%) 786981.254 ns/op [Average]
[info]   (min, avg, max) = (77933768.721, 78726787.768, 79423429.968), stdev = 520539.373
[info]   CI (99.9%): [77939806.514, 79513769.022] (assumes normal distribution)
[info] # Run complete. Total time: 00:03:21
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                  Mode  Cnt         Score        Error  Units
[info] GoogleThinkBenchmark.testAesGcmJceEncrypt  avgt   10  78726787.768 ± 786981.254  ns/op
```
with 
"com.google.crypto.tink" % "tink" % "1.6.1"
```
[info] compiling 5 Java sources to /home/bjorn/Documents/github/GoogleThinkBenchmark/target/scala-2.12/classes ...
[info] running (fork) org.openjdk.jmh.Main -i 10 -wi 10 -f 1 -t 1
[info] # JMH version: 1.31
[info] # VM version: JDK 17.0.6, OpenJDK 64-Bit Server VM, 17.0.6+10-Ubuntu-1ubuntu2
[info] # VM invoker: /usr/lib/jvm/java-17-openjdk-amd64/bin/java
[info] # VM options: <none>
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 10 iterations, 10 s each
[info] # Measurement: 10 iterations, 10 s each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread, will synchronize iterations
[info] # Benchmark mode: Average time, time/op
[info] # Benchmark: com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt
[info] # Run progress: 0.00% complete, ETA 00:03:20
[info] # Fork: 1 of 1
[info] # Warmup Iteration   1: 83260488.298 ns/op
[info] # Warmup Iteration   2: 77947987.496 ns/op
[info] # Warmup Iteration   3: 78086226.109 ns/op
[info] # Warmup Iteration   4: 76767415.580 ns/op
[info] # Warmup Iteration   5: 76449646.229 ns/op
[info] # Warmup Iteration   6: 78385935.750 ns/op
[info] # Warmup Iteration   7: 77999410.016 ns/op
[info] # Warmup Iteration   8: 76971635.577 ns/op
[info] # Warmup Iteration   9: 76934072.685 ns/op
[info] # Warmup Iteration  10: 77180919.354 ns/op
[info] Iteration   1: 76979372.431 ns/op
[info] Iteration   2: 76777254.260 ns/op
[info] Iteration   3: 76748232.038 ns/op
[info] Iteration   4: 77181769.492 ns/op
[info] Iteration   5: 78160841.664 ns/op
[info] Iteration   6: 77396535.208 ns/op
[info] Iteration   7: 77519909.492 ns/op
[info] Iteration   8: 77570303.233 ns/op
[info] Iteration   9: 77661022.992 ns/op
[info] Iteration  10: 78231386.313 ns/op
[info] Result "com.googlethinkbenchmark.GoogleThinkBenchmark.testAesGcmJceEncrypt":
[info]   77422662.712 ±(99.9%) 782935.090 ns/op [Average]
[info]   (min, avg, max) = (76748232.038, 77422662.712, 78231386.313), stdev = 517863.085
[info]   CI (99.9%): [76639727.623, 78205597.802] (assumes normal distribution)
[info] # Run complete. Total time: 00:03:21
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                                  Mode  Cnt         Score        Error  Units
[info] GoogleThinkBenchmark.testAesGcmJceEncrypt  avgt   10  77422662.712 ± 782935.090  ns/op
[success] Total time: 207 s (03:27), completed May 7, 2023, 9:13:15 PM
```

Almost the same.. Think 1.9.0 is perhaps a bit slower.

**Comment by dongjoon-hyun** [comment]
Could you put that into the PR description in order to make a commit log, @bjornjorgensen ?

**Comment by bjornjorgensen** [comment]
ok, I have added 
This have be benchmarks tested 
With
"com.google.crypto.tink" % "tink" % "1.6.1"
(min, avg, max) = (75024163.500, 76331532.832, 77324718.069), stdev = 652319.870

With
"com.google.crypto.tink" % "tink" % "1.9.0"
(min, avg, max) = (76279051.841, 77512667.749, 78590966.453), stdev = 632832.384

Almost the same.. Think 1.9.0 is perhaps a bit slower.

Is it ok or?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.5.0.
Thank you, @bjornjorgensen and @LuciferYang .

**Comment by LuciferYang** [comment]
late LGTM

**Pull request closed without merging**